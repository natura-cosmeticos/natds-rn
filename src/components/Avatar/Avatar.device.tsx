import { storiesOf } from '@storybook/react-native';
import {
  all,
  interactive,
} from './Avatar.stories';

storiesOf('Avatar', module)
  .add('all', all)
  .add('interactive', interactive);
