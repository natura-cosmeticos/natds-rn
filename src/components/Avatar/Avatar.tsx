/* eslint-disable max-lines */
import React from 'react';
import {
  AvatarIcon, AvatarImage, AvatarLetter, Container,
} from './Avatar.styles';
import {
  AvatarIconProps,
  AvatarImageProps, AvatarLetterProps, AvatarProps,
} from './Avatar.types';
// @ts-ignore
import IconAnonymous from '../../assets/images/anonymous.jpg';

/**
 * Get the first character from first and last word
 * @param text string
 */
export const getTextValue = (text = '') => {
  const textFormated = text.trim().toUpperCase();

  const firstLetters = textFormated
    .trim()
    .split(' ')
    .map(item => item.charAt(0))
    .join();

  if (firstLetters.length < 2) return firstLetters;

  return `${firstLetters.charAt(0)}${firstLetters.charAt(firstLetters.length - 1)}`;
};

export const isAvatarLetter = (props: AvatarProps): props is AvatarLetterProps => props.type === 'letter';
export const isAvatarImage = (props: AvatarProps): props is AvatarImageProps => props.type === 'image';
export const isAvatarIcon = (props: AvatarProps): props is AvatarIconProps => props.type === 'icon';

export const Avatar = (props: AvatarProps) => {
  const {
    size = 'standard',
    testID = 'avatar',
    accessibilityLabel,
    accessibilityHint,
    type = 'anonymous',
  } = props;

  return (
    <Container size={size}>
      {isAvatarLetter(props) && (
        <AvatarLetter
          accessibilityLabel={accessibilityLabel}
          accessibilityHint={accessibilityHint}
          accessibilityRole="text"
          testID={testID}
          size={size}
          type={type}
        >
          {getTextValue(props.text)}
        </AvatarLetter>
      )}
      {isAvatarImage(props) && (
        <AvatarImage
          accessibilityLabel={accessibilityLabel}
          accessibilityHint={accessibilityHint}
          accessibilityRole="image"
          testID={testID}
          size={size}
          type={type}
          source={props.imgSource}
        />
      )}
      {isAvatarIcon(props) && (
        <AvatarIcon
          accessibilityHint={accessibilityHint}
          accessibilityLabel={accessibilityLabel}
          accessibilityRole="image"
          testID={testID}
          size={size}
          name={props.iconName}
        />
      )}
      {type === 'anonymous' && (
        <AvatarImage
          accessibilityLabel={accessibilityLabel}
          accessibilityHint={accessibilityHint}
          accessibilityRole="image"
          testID={testID}
          size={size}
          type={type}
          source={IconAnonymous}
        />
      )}
    </Container>
  );
};
