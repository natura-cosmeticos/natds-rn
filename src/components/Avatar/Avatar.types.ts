/* eslint-disable max-len */
import { IconName } from '@naturacosmeticos/natds-icons'
import { Size } from '@naturacosmeticos/natds-themes/react-native'

/**
 * @deprecated
 */
type AvatarSizesDeprecated = keyof Pick<Size, 'tiny' | 'small' | 'large' | 'huge'>;
export type AvatarSizes = keyof Pick<Size, 'standard' | 'semi' | 'semiX' | 'medium' | 'largeXXX' | AvatarSizesDeprecated>;
/**
 * @deprecated
 */
type AvatarTypesDeprecated = 'anonymous';
export type AvatarTypes = 'image' | 'icon' | 'letter' | AvatarTypesDeprecated;

type AvatarUriProps = {
  uri: string;
}

export interface AvatarBaseProps {
  /**
   * Optional size of the avatar.
   * Deprecated `tiny`, `small`, `large` and `huge` sizes.
   */
  size?: AvatarSizes;
  /**
   * Optional accessibilityLabel.
   */
  accessibilityLabel?: string;
  /**
   * Optional accessibilityHint.
   */
  accessibilityHint?: string;
  /**
   * Optional ID for testing.
   */
  testID?: string;
}

export type AvatarAnonymousProps = AvatarBaseProps & {
  type?: 'anonymous';
}

export type AvatarImageProps = AvatarBaseProps & {
  type?: 'image';
  /**
  * URL image.
  * @example imgSource={{uri: IMG_URL}}
  */
  imgSource?: string | AvatarUriProps;
}

export type AvatarIconProps = AvatarBaseProps & {
  type?: 'icon';
  /**
  * Icon name
  */
  iconName?: IconName;
}

export type AvatarLetterProps = AvatarBaseProps & {
  type?: 'letter';
  /**
  * Label text
  */
  text?: string;
}

export type AvatarProps = AvatarImageProps | AvatarIconProps | AvatarLetterProps | AvatarAnonymousProps;
