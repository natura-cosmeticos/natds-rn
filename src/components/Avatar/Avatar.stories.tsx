import React from 'react';
import { select, text } from '@storybook/addon-knobs';
import { IconName, iconNames } from '@naturacosmeticos/natds-icons';

import { AvatarTypes, AvatarSizes } from './Avatar.types';
// @ts-ignore
import NatAvatar from '../../assets/images/nat_avatar.jpg';
import { StoryContainer } from '../../common/HelperComponents/StoryContainer';
import { Avatar } from '.';

const description = () => `
---

**NOTE**: This component is available in the following types:

  - ✅ icon
  - ✅ image
  - ✅ letter

With the following attribute status:

- **Size:**
  - ✅ \`standard\`
  - ✅ \`semi\`
  - ✅ \`semiX\`
  - ✅ \`medium\`
  - ✅ \`largeXXX\`
- ✅ **imgSource**  (available only in image type)
- ✅ **text**  (available only in letter type)
- ✅ **iconName** (available only in icon type)

---
`;

export default {
  component: Avatar,
  parameters: {
    componentSubtitle: 'Avatars are used to identify a user.',
    docs: {
      extractComponentDescription: description,
    },
  },
  title: 'Components/Avatar',
};

const textSample = 'Design System';
const iconSample = 'outlined-default-mockup';

const avatarTypes = {
  icon: 'icon',
  image: 'image',
  letter: 'letter',
};

const avatarSizes = {
  largeXXX: 'largeXXX',
  medium: 'medium',
  semi: 'semi',
  semiX: 'semiX',
  standard: 'standard',
};

export const All = () => (
  <>
    <StoryContainer title="Image">
      <Avatar type="image" size="standard" imgSource={NatAvatar}/>
      <Avatar type="image" size="semi" imgSource={NatAvatar}/>
      <Avatar type="image" size="semiX" imgSource={NatAvatar}/>
      <Avatar type="image" size="medium" imgSource={NatAvatar}/>
      <Avatar type="image" size="largeXXX" imgSource={NatAvatar}/>
    </StoryContainer>
    <StoryContainer title="Letter">
      <Avatar type="letter" size="standard" text={textSample}/>
      <Avatar type="letter" size="semi" text={textSample}/>
      <Avatar type="letter" size="semiX" text={textSample}/>
      <Avatar type="letter" size="medium" text={textSample}/>
      <Avatar type="letter" size="largeXXX" text={textSample}/>
    </StoryContainer>
    <StoryContainer title="Icon">
      <Avatar type="icon" size="standard" iconName={iconSample}/>
      <Avatar type="icon" size="semi" iconName={iconSample}/>
      <Avatar type="icon" size="semiX" iconName={iconSample}/>
      <Avatar type="icon" size="medium" iconName={iconSample}/>
      <Avatar type="icon" size="largeXXX" iconName={iconSample}/>
    </StoryContainer>
  </>
);

export const Interactive = () => (
  <StoryContainer title="Interactive">
    <Avatar
      type={select('Type', avatarTypes, 'icon') as AvatarTypes}
      text={text('Text', 'Natura Design System')}
      imgSource={text('URL', '')}
      iconName={select('IconName', iconNames as Array<IconName>, 'outlined-default-mockup' as IconName)}
      size={select('Size', avatarSizes, 'medium') as AvatarSizes}
    />
  </StoryContainer>
);
