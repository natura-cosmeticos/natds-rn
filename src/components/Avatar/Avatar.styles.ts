/* eslint-disable max-len */
import styled from 'styled-components/native';
import {
  Theme, getColorOnPrimary, getColorPrimary, getAvatarBySize,
} from '../../common/themeSelectors';
import { AvatarSizes } from './Avatar.types';
import { Icon } from '../Icon/Icon';

interface AvatarStyleProps {
  size: AvatarSizes;
  theme: Theme;
}


// List of deprecated sizes used to conditionally render the avatar
const deprecatedSizes = ['tiny', 'small', 'large', 'huge'];

const getAvatarFontSize = (theme, size) => getAvatarBySize(size, theme).fontSize;

const getAvatarSize = (theme, size) => getAvatarBySize(size, theme).size;

const getBorderRadiusBySize = (size: AvatarSizes, theme: Theme) => {
  if (!deprecatedSizes.includes(size)) {
    return {
      borderRadius: theme.avatar[size].borderRadius,
      height: theme.size[size],
      width: theme.size[size],
    };
  }

  return {
    borderRadius: getAvatarSize(theme, size),
    height: getAvatarSize(theme, size),
    width: getAvatarSize(theme, size),
  };
};

const getFontsBySize = (size: AvatarSizes, theme: Theme) => {
  if (!deprecatedSizes.includes(size)) {
    return {
      fontFamily: theme.avatar.primary.fontFamily,
      fontSize: theme.avatar[size].fontSize,
      fontWeight: theme.avatar.primary.fontWeight,
      letterSpacing: theme.avatar[size].letterSpacing,
      lineHeight: theme.avatar[size].fontSize * theme.avatar[size].lineHeight,
    };
  }

  return { fontSize: getAvatarFontSize(theme, size) };
};

export const Container = styled.View(({ size, theme }: AvatarStyleProps) => ({
  ...getBorderRadiusBySize(size, theme),
  alignItems: 'center',
  backgroundColor: getColorPrimary(theme),
  justifyContent: 'center',
}));

export const AvatarImage = styled.Image(({ size, theme }: AvatarStyleProps) => ({
  ...getBorderRadiusBySize(size, theme),
}));

export const AvatarLetter = styled.Text(({ size, theme }: AvatarStyleProps) => ({
  ...getFontsBySize(size, theme),
  alignSelf: 'center',
  color: getColorOnPrimary(theme),
  fontFamily: theme.avatar.primary.fontFamily,
  fontWeight: theme.avatar.primary.fontWeight,

}));

export const AvatarIcon = styled(Icon)(({ theme }) => ({
  color: getColorOnPrimary(theme),
}));
