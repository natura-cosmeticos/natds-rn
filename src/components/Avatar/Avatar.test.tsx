/* eslint-disable max-statements */
import React from 'react';

import { Avatar, getTextValue } from './Avatar';
import { renderWithTheme } from '../../../test/testHelpers';

jest.mock('../Icon/Icon');
jest.mock('../../common/themeSelectors', () => (
  {
    getAvatarBySize: () => ({ fontSize: 24, size: 'largeXX' }),
    getColorOnPrimary: () => '#333333',
    getColorPrimary: () => '#FFAA33',
  }));

describe('Avatar component', () => {
  it('should render avatar with default size', () => {
    const { queryByTestId } = renderWithTheme(<Avatar testID='ds-avatar' type='image' imgSource={{ uri: '' }} />);

    expect(queryByTestId('ds-avatar')).toHaveStyle({ height: 24, width: 24 });
  });

  it('should render avatar with image type', () => {
    const { queryByTestId } = renderWithTheme(<Avatar testID='ds-avatar-image' type='image' imgSource={{ uri: '' }} />);

    expect(queryByTestId('ds-avatar-image')?.props).toHaveProperty('type', 'image');
  });

  it('should render avatar with letter type', () => {
    const { queryByTestId } = renderWithTheme(<Avatar testID='ds-avatar-letter' type='letter' text="Design System" />);

    expect(queryByTestId('ds-avatar-letter')?.props).toHaveProperty('type', 'letter');
  });

  it('should render avatar component icon', () => {
    const { toJSON } = renderWithTheme(<Avatar type='icon' iconName='filled-default-mockup' />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render avatar component image', () => {
    const { toJSON } = renderWithTheme(<Avatar type='image' imgSource={{ uri: '' }} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render avatar component letter', () => {
    const { toJSON } = renderWithTheme(<Avatar type='letter' text='Design System' />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render avatar component anonymoys', () => {
    const { toJSON } = renderWithTheme(<Avatar />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render avatar with a deprecated size', () => {
    const { toJSON, getByTestId } = renderWithTheme(<Avatar testID='avatar-letter' type='letter' size='huge' text='Deprecated'/>);

    expect(getByTestId('avatar-letter')).toHaveStyle({ fontSize: 24 });
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render avatar component with other size', () => {
    const { queryByTestId } = renderWithTheme(<Avatar type='image' imgSource={{ uri: '' }} size='semiX' />);

    expect(queryByTestId('avatar')).toHaveStyle({ height: 40, width: 40 });
  });

  it('should return the first character from first and last word', () => {
    const mockFn = jest.fn(getTextValue);

    mockFn();

    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it('should return the first character', () => {
    const mockFn = jest.fn(getTextValue);

    expect(mockFn('Design')).toBe('D');
  });
});
