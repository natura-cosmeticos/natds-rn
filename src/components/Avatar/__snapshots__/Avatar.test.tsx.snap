// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar component should render avatar component anonymoys 1`] = `
<View
  size="standard"
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFAA33",
        "borderRadius": 12,
        "height": 24,
        "justifyContent": "center",
        "width": 24,
      },
    ]
  }
>
  <Image
    accessibilityRole="image"
    size="standard"
    source={
      Object {
        "testUri": "../../../src/assets/images/anonymous.jpg",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 12,
          "height": 24,
          "width": 24,
        },
      ]
    }
    testID="avatar"
    type="anonymous"
  />
</View>
`;

exports[`Avatar component should render avatar component icon 1`] = `
<View
  size="standard"
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFAA33",
        "borderRadius": 12,
        "height": 24,
        "justifyContent": "center",
        "width": 24,
      },
    ]
  }
>
  <Text
    accessibilityRole="image"
    name="filled-default-mockup"
    size="standard"
    style={
      Array [
        Object {
          "color": "#333333",
        },
      ]
    }
    testID="avatar"
  >
    filled-default-mockup
  </Text>
</View>
`;

exports[`Avatar component should render avatar component image 1`] = `
<View
  size="standard"
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFAA33",
        "borderRadius": 12,
        "height": 24,
        "justifyContent": "center",
        "width": 24,
      },
    ]
  }
>
  <Image
    accessibilityRole="image"
    size="standard"
    source={
      Object {
        "uri": "",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 12,
          "height": 24,
          "width": 24,
        },
      ]
    }
    testID="avatar"
    type="image"
  />
</View>
`;

exports[`Avatar component should render avatar component letter 1`] = `
<View
  size="standard"
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFAA33",
        "borderRadius": 12,
        "height": 24,
        "justifyContent": "center",
        "width": 24,
      },
    ]
  }
>
  <Text
    accessibilityRole="text"
    size="standard"
    style={
      Array [
        Object {
          "alignSelf": "center",
          "color": "#333333",
          "fontFamily": "Roboto",
          "fontSize": 12,
          "fontWeight": "400",
          "letterSpacing": 0,
          "lineHeight": 18,
        },
      ]
    }
    testID="avatar"
    type="letter"
  >
    DS
  </Text>
</View>
`;

exports[`Avatar component should render avatar with a deprecated size 1`] = `
<View
  size="huge"
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFAA33",
        "borderRadius": "largeXX",
        "height": "largeXX",
        "justifyContent": "center",
        "width": "largeXX",
      },
    ]
  }
>
  <Text
    accessibilityRole="text"
    size="huge"
    style={
      Array [
        Object {
          "alignSelf": "center",
          "color": "#333333",
          "fontFamily": "Roboto",
          "fontSize": 24,
          "fontWeight": "400",
        },
      ]
    }
    testID="avatar-letter"
    type="letter"
  >
    D
  </Text>
</View>
`;
